libraries/AP_DAL/AP_DAL_GPS.cpp

```diff
@@ -0,0 +1,55 @@
+#include "AP_DAL_GPS.h"
+
+#include <AP_Logger/AP_Logger.h>
+#include "AP_DAL.h"
+
+AP_DAL_GPS::AP_DAL_GPS()
+{
+    for (uint8_t i=0; i<ARRAY_SIZE(_RGPI); i++) {
+        _RGPI[i].instance = i;
+        _RGPJ[i].instance = i;
+    }
+}
+
+void AP_DAL_GPS::start_frame()
+{
+    const auto &gps = AP::gps();
+
+    const log_RGPH old_RGPH = _RGPH;
+    _RGPH.primary_sensor = gps.primary_sensor();
+    _RGPH.num_sensors = gps.num_sensors();
+
+    WRITE_REPLAY_BLOCK_IFCHANGD(RGPH, _RGPH, old_RGPH);
+
+    for (uint8_t i=0; i<ARRAY_SIZE(_RGPI); i++) {
+        log_RGPI &RGPI = _RGPI[i];
+        const log_RGPI old_RGPI = RGPI;
+
+        RGPI.status = (GPS_Status)gps.status(i);
+
+        const Location &loc = gps.location(i);
+        RGPI.last_message_time_ms = gps.last_message_time_ms(i);
```

Comment:
```suggestion
        RGPI.last_message_time_ms = gps.last_message_time_ms(i);
        const Location &loc = gps.location(i);
```

(just swapping for clarity)